File: frontend/src/api/ingestion.ts
import { api } from "./client";

export interface IngestionRunOut {
  id: string;
  source_file: string;
  chunks_written: number;
  chunks_skipped: number;
  chunks_deleted: number;
  embedding_model_version: string;
  started_at: string;
  completed_at: string | null;
  status: string;
  notes: string | null;
}

export interface IngestionStatusOut {
  collection_counts: Record<string, number>;
  last_run_at: string | null;
}

export const ingestionApi = {
  run: (sources?: string[]) => api.post<IngestionRunOut[]>("/ingestion/run", { sources: sources ?? null }),
  runs: () => api.get<IngestionRunOut[]>("/ingestion/runs"),
  status: () => api.get<IngestionStatusOut>("/ingestion/status"),
};

File: backend/app/api/ingestion.py
from fastapi import APIRouter
from pydantic import BaseModel
from sqlalchemy import select

from app.deps import SessionDep
from app.ingestion.index_pipeline import run_curriculum_ingestion, run_skkni_ingestion
from app.models import IngestionRun
from app.schemas import IngestionRunOut, IngestionStatusOut
from app.vectorstore import curriculum_collection, skkni_collection

router = APIRouter(prefix="/ingestion", tags=["ingestion"])


class IngestionRunRequest(BaseModel):
    sources: list[str] | None = None  # any of "skkni", "curriculum"; None runs both


@router.post("/run", response_model=list[IngestionRunOut])
async def trigger_ingestion(payload: IngestionRunRequest, session: SessionDep):
    sources = payload.sources or ["skkni", "curriculum"]
    runs = []
    if "skkni" in sources:
        runs.append(await run_skkni_ingestion(session))
    if "curriculum" in sources:
        runs.append(await run_curriculum_ingestion(session))
    return runs


@router.get("/runs", response_model=list[IngestionRunOut])
async def list_ingestion_runs(session: SessionDep):
    runs = (await session.execute(select(IngestionRun).order_by(IngestionRun.started_at.desc()))).scalars().all()
    return runs


@router.get("/status", response_model=IngestionStatusOut)
async def ingestion_status(session: SessionDep):
    last_run = (
        await session.execute(select(IngestionRun).order_by(IngestionRun.started_at.desc()).limit(1))
    ).scalar_one_or_none()
    return IngestionStatusOut(
        collection_counts={
            "skkni_units": await skkni_collection().count(),
            "curriculum_courses": await curriculum_collection().count(),
        },
        last_run_at=last_run.started_at if last_run else None,
    )
